Civi/Core/Url.php

```diff
@@ -0,0 +1,353 @@
+          break;
+
+        // (r)elative url
+        case 'r':
+          $this->preferFormat = 'relative';
+          break;
+
+        // (h)tml encoding
+        case 'h':
+          $this->htmlEscape = TRUE;
+          break;
+
+        // (p)lain text encoding
+        case 'p':
+          $this->htmlEscape = FALSE;
+          break;
+
+        // (s)sl
+        case 's';
+          $this->ssl = TRUE;
+          break;
+      }
+    }
+    return $this;
+  }
+
+  /**
+   * Render the final URL as a string.
+   *
+   * @return string
+   */
+  public function __toString(): string {
+    $userSystem = \CRM_Core_Config::singleton()->userSystem;
+    $scheme = $this->getScheme();
+    $preferFormat = $this->getPreferFormat();
+
+    // Translate subjective values to real values.
+    switch ($scheme) {
+      case 'current':
+        $preferFormat = $preferFormat ?: 'relative';
+        $scheme = $userSystem->isFrontEndPage() ? 'frontend' : 'backend';
+        // The current call could actually be a 'service' request, but we treat those as equivalent to 'frontend', so maybe it doesn't matter.
+        break;
+
+      case 'default':
+        // $preferFormat = $preferFormat ?: 'absolute';
+        // TODO pick $scheme = 'frontend' or 'backend' or 'service';
+        throw new \RuntimeException("FIXME: Implement lookup for default ");
+
+      default:
+        $preferFormat = $preferFormat ?: 'absolute';
+    }
+
+    switch ($scheme) {
+      case 'frontend':
+      case 'service':
+        $result = $userSystem->url($this->getPath(), $this->getQuery(), $preferFormat === 'absolute', $this->getFragment(), TRUE, FALSE, FALSE);
+        break;
+
+      case 'backend':
+        $result = $userSystem->url($this->getPath(), $this->getQuery(), $preferFormat === 'absolute', $this->getFragment(), FALSE, TRUE, FALSE);
+        break;
+
+      case 'assetBuilder':
+        $assetName = $this->getPath();
+        $assetParams = [];
+        parse_str('' . $this->getQuery(), $assetParams);
+        $result = \Civi::service('asset_builder')->getUrl($assetName, $assetParams);
+        break;
+
+      case 'ext':
+        $parts = explode('/', $this->getPath(), 2);
+        $result = \Civi::resources()->getUrl($parts[0], $parts[1] ?? NULL, FALSE);
+        if ($this->query) {
+          $result .= '?' . $this->query;
+        }
+        if ($this->fragment) {
+          $result .= '#' . $this->fragment;
+        }
+        break;
+
+      default:
+        throw new \RuntimeException("Unknown URL scheme: {$this->getScheme()}");
+    }
+
+    // TODO decide if the current default is good enough for future
+    $ssl = $this->getSsl() ?: \CRM_Utils_System::isSSL();
+    if ($ssl && str_starts_with($result, 'http:')) {
+      $result = 'https:' . substr($result, 5);
+    }
+    elseif (!$ssl && str_starts_with($result, 'https:')) {
+      $result = 'http:' . substr($result, 6);
+    }
+
+    return $this->htmlEscape ? htmlentities($result) : $result;
```

Comment:
But seriously, can we just not include this oddball feature in the initial implementation and hold off on adding it unless some pressing reason emerged in the future. I'd much rather have a new function be simpler and missing a dumb feature than more complex with yagni baggage.

Reply:
@colemanw Right, `$htmlize` has been pretty annoying-surprising. 

My hesitation for completely removing it:

* At some point, I expect we'd start wanting to convert existing calls (`CRM_Utils_System::url()` => `Civi::url()`). I bet 80% of those conversions are easier if you still have a flag for HTML.
* When encoding some data to go into an HTML document, there are different flavors (`htmlentities()`, `htmlspecialchars()`, `htmlspecialchars(...,ENT_QUOTES)`). One of those three is more susceptible to bugs (*re: CSRF, i18n, etc*) than the others. How often will people correctly remember which one?
* Just general gut sense that it's deeply ingrained, and that we'll bump into weird edges in removing it.

There's a sort of mirror-image between the PHP and Smarty (`{crmURL...}`) generation of URLs:

* In PHP, you should probably get the canonical URL most of the time -- but sometimes you want the HTML-encoded variant.
* In Smarty, you should probably get the HTML-encoded URL most of the time -- but sometimes you might another format (e.g. for JS-in-Smarty).

So at high level,  PHP and Smarty need different defaults but also need ways to flip over. It's kind of appealing that you have same list of allowed `$flags` regardless of the medium. Though TBH, I haven't really thought about the Smarty POV.

Reply:
> I bet 80% of those conversions are easier if you still have a flag for HTML

Well... In a proper MVC, the escaping is done by the template layer. Definitely the smarty implementation ought to run the output of this function through `htmlentities()`, but it doesn't need a flag on this function to do that, it can just do it.
Angular pages definitely *don't* need the escaping, and neither do internal calls to something like `pushUserContext` or `redirect`. I can see that a lot of calls to `CRM_Utils_System::url()` accidentally leave escaping on when they shouldn't; for example (not to pick on anyone on purpose this is just the first one I found) @eileen just added one in 2023 that does escaping when it shouldn't:

https://github.com/civicrm/civicrm-core/blob/8ec2d293af07a1f2b65173a86a9afcd2488d32c1/CRM/Core/Form.php#L2711

I'm sure I've made that same mistake a dozen times or more. It's such a prevalent problem that we even have this aptly-named function:

https://github.com/civicrm/civicrm-core/blob/86d51faf4cfa430b2178b20c0c80fcbeafa41efc/CRM/Utils/String.php#L777-L790

So even if you're right and 80% of calls to `CRM_Utils_System::url` use escaping, I would hazard that probably 80% of those, *shouldn't*.

So what to do? Either we leave the flag in with a "this is dumb" warning, or take it out. I can live with it either way, as long as we don't ever attempt an auto-conversion of `CRM_Utils_System::url(...)` => `Civi::url()` that blindly copies over that flag. Every usage of `CRM_Utils_System` with escaping *on* is smelly and needs to be audited IMO.

Reply:
While I was poking in there I noticed something really nutty: https://github.com/civicrm/civicrm-core/pull/26888